themes/jquery/header.php

```diff
@@ -1,11 +1,11 @@
+<?php jq_content_security_policy() ?>
```

Comment:
I'm not sure this is safe. These template files are specifically for HTML output.

WordPress provides two action hooks for sending headers that we can use instead:

* https://developer.wordpress.org/reference/hooks/send_headers/
* https://developer.wordpress.org/reference/hooks/wp_headers/

Between these, I'd recommend `send_headers`. It has the benefit of only being sent on HTTP 304 responses, where sending a new copy (but without a new body) would cause the two to get out of sync if anything uses the `$nonce`.

Popular plugins like WordPress "http-headers" also use this hook.

Reply:
Fair enough, but I don't think it's unsafe. A header can only be set before any HTML output. But, I like `send_headers`.

Reply:
I've not verified the following, but I believe WordPress can buffer the "template" (theme) output separately from headers, and thus has control over when and if theme-based headers, and when theme HTTP body is sent. Emitting a header directly from inside a template file thus bypasses that, which I would describe as unsafe/undocumented. This can manifest for example when rendering an error page. If you ob_start and then ob_flush or end due to an uncaught error, you cleanly undo the output buffer, yet you'd be left with headers. That's "fine" in the case of CSP since we don't mind it being on an error response, but it's one of those things that may bite you in the future when debugging something or after an automatic WordPress upgrade.

It might also cause unexpected side-effects with caching plugins, those can trigger background jobs that can safely and proactively generate page output from a template (and then trigger headers "live" on the cache serving), but if those template calls already trigger header calls, those may end up causing errors in background jobs or sent to on unrelated wp-admin requests that wasn't expected. Plus, they'd be missing from cache servings.

I trust you've tested this and it probably doesn't affect us, but with unattended and stress-free upgrades, I kinda like not having to worry about how undocumented behaviours might affect us in the future, even if (admitedly) fairly unlikely for our setup.

Reply:
I don't know if I'd call it undocumented. It's not a wordpress function. It's part of [php](https://www.php.net/manual/en/function.header.php). And because it can only be placed before the first HTML output (out of all templates), I think it's kept separate from the buffer you're talking about.

Still, I like the separation of concerns and will switch it to `send_headers`.

Reply:
Now using `send_headers`